x-pack/plugins/aiops/public/components/log_categorization/attachments_menu.tsx

```diff
@@ -0,0 +1,247 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { EuiContextMenuProps } from '@elastic/eui';
+import {
+  EuiButton,
+  EuiButtonIcon,
+  EuiContextMenu,
+  EuiFlexItem,
+  EuiForm,
+  EuiFormRow,
+  EuiPanel,
+  EuiPopover,
+  EuiSpacer,
+  EuiSwitch,
+} from '@elastic/eui';
+import type { CasesPublicStart } from '@kbn/cases-plugin/public';
+import type { Capabilities } from '@kbn/core-capabilities-common';
+import type { EmbeddableStart } from '@kbn/embeddable-plugin/public';
+import { i18n } from '@kbn/i18n';
+import { FormattedMessage } from '@kbn/i18n-react';
+import type { SaveModalDashboardProps } from '@kbn/presentation-util-plugin/public';
+import {
+  LazySavedObjectSaveModalDashboard,
+  withSuspense,
+} from '@kbn/presentation-util-plugin/public';
+import React, { useCallback, useState } from 'react';
+import { useMemo } from 'react';
+import type { DataView } from '@kbn/data-views-plugin/common';
+import { EMBEDDABLE_PATTERN_ANALYSIS_TYPE } from '@kbn/aiops-log-pattern-analysis/constants';
+import { useTimeRangeUpdates } from '@kbn/ml-date-picker';
+import type { PatternAnalysisEmbeddableState } from '../../embeddables/pattern_analysis/types';
+import type { RandomSamplerOption, RandomSamplerProbability } from './sampling_menu/random_sampler';
+import { useCasesModal } from '../../hooks/use_cases_modal';
+
+const SavedObjectSaveModalDashboard = withSuspense(LazySavedObjectSaveModalDashboard);
+
+interface AttachmentsMenuProps {
+  dataView: DataView;
+  randomSamplerMode: RandomSamplerOption;
+  randomSamplerProbability: RandomSamplerProbability;
+  embeddable?: EmbeddableStart;
+  cases?: CasesPublicStart;
+  capabilities?: Capabilities;
+  selectedField?: string;
+}
+
+export const AttachmentsMenu = ({
+  embeddable,
+  cases,
+  capabilities,
+  dataView,
+  selectedField,
+  randomSamplerMode,
+  randomSamplerProbability,
+}: AttachmentsMenuProps) => {
+  const [applyTimeRange, setApplyTimeRange] = useState(false);
+
+  const [dashboardAttachmentReady, setDashboardAttachmentReady] = useState(false);
+  const [isActionMenuOpen, setIsActionMenuOpen] = useState(false);
+
+  const { create: canCreateCase, update: canUpdateCase } = cases?.helpers?.canUseCases() ?? {
+    create: false,
+    update: false,
+  };
+
+  const openCasesModalCallback = useCasesModal(EMBEDDABLE_PATTERN_ANALYSIS_TYPE);
+
+  const timeRange = useTimeRangeUpdates();
+
+  const canEditDashboards = capabilities?.dashboard?.createNew ?? false;
+
+  const onSave: SaveModalDashboardProps['onSave'] = useCallback(
+    ({ dashboardId, newTitle, newDescription }) => {
+      const stateTransfer = embeddable!.getStateTransfer();
```

Comment:
Nit: Is it possible to avoid the !? If you keep passing in embeddable via props you could make it non-optional, other maybe return early? (I saw it's the same for the change point one so probably fine to leave as is if too much trouble refactoring)

Reply:
The embeddable is no longer passed as a prop since we're now calling `useAiopsAppContext` directly. In this scenario, I don't think there's much we can do about it. We could return early in the onSave callback, but that’s all.
I tried modifying the AiopsAppContext type to differentiate when embeddable should be defined, but achieving this isn’t that easy.